### PR TITLE
[ALLUXIO-2050] Add an option to force loading metadata

### PR DESCRIPTION
**Pull request by peisun1115**
https://alluxio.atlassian.net/browse/ALLUXIO-2050

### CODE REVIEW

**Comment by peisun1115** [comment]
@jsimsa @gpang

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 1]
```diff
@@ -13,6 +13,7 @@
 
 import alluxio.annotation.PublicApi;
 import alluxio.thrift.ListStatusTOptions;
+import alluxio.thrift.LoadMetadataType;
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: the thrift type should not be exposed to the application layer, which it currently is ... we need separate classes for Alluxio LoadMetadataType and Thrift LoadMetadataType

Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: (exposing Thrift types to the application layer prevents us from replacing Thrift with other network protocol without breaking backwards compatibility)

File: core/common/src/thrift/common.thrift [hunk 2]
```diff
@@ -28,6 +28,12 @@ enum CommandType {
   Persist = 5,  // Ask the worker to persist a file for lineage
 }
 
+enum LoadMetadataType {
```
Comment (jsimsa) [inline core/common/src/thrift/common.thrift]: move this to `file_system_master.thrift`

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 3]
```diff
@@ -510,15 +512,14 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
...
-   *        direct children have not beed loaded before
+   * @param getFileInfoListOptions the {@link GetFileInfoListOptions}
    * @return the list of {@link FileInfo}s
    * @throws AccessControlException if permission checking fails
    * @throws FileDoesNotExistException if the file does not exist
    * @throws InvalidPathException if the path is invalid
    */
-  // TODO(peis): Create GetFileInfoListOptions and add an option not to load metadata.
-  public List<FileInfo> getFileInfoList(AlluxioURI path, boolean loadDirectChildren)
+  public List<FileInfo> getFileInfoList(AlluxioURI path,
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: while we are doing this, can we rename this function to `listStatus` and its options to ListStatusOptions? thanks

File: core/server/src/main/java/alluxio/master/file/FileSystemMasterClientRestServiceHandler.java [hunk 4]
```diff
@@ -251,20 +255,46 @@ public Response free(@QueryParam("path") String path,
...
+   * removed in 2.0. Use LIST_STATUS2.
    * @param path the file path
    * @param loadDirectChildren whether to load direct children of path
    * @return the response object
+   * @deprecated since 1.1.1 and will be removed in 2.0
    */
   @GET
   @Path(LIST_STATUS)
   @ReturnType("java.util.List<alluxio.wire.FileInfo>")
+  @Deprecated
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/FileSystemMasterClientRestServiceHandler.java]: adding a param here is backwards compatible, so no need to create a new endpoint

File: core/server/src/main/java/alluxio/master/file/FileSystemMasterClientRestServiceHandler.java [hunk 5]
```diff
@@ -19,8 +19,10 @@
 import alluxio.master.file.options.CompleteFileOptions;
 import alluxio.master.file.options.CreateDirectoryOptions;
 import alluxio.master.file.options.CreateFileOptions;
+import alluxio.master.file.options.GetFileInfoListOptions;
 import alluxio.master.file.options.MountOptions;
 import alluxio.master.file.options.SetAttributeOptions;
+import alluxio.thrift.LoadMetadataType;
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/FileSystemMasterClientRestServiceHandler.java]: this class should be using Alluxio `LoadMetadataType`

File: tests/src/test/java/alluxio/master/file/FileSystemMasterClientRestApiTest.java [hunk 6]
```diff
@@ -179,14 +181,37 @@ public void listStatusTest() throws Exception {
...
     new TestCase(mHostname, mPort,
         getEndpoint(FileSystemMasterClientRestServiceHandler.LIST_STATUS), params, HttpMethod.GET,
         fileInfos).run();
 
     Mockito.verify(mFileSystemMaster)
-        .getFileInfoList(Mockito.<AlluxioURI>any(), Mockito.anyBoolean());
+        .getFileInfoList(Mockito.<AlluxioURI>any(), Mockito.<GetFileInfoListOptions>any());
+  }
+
+  @Test
```
Comment (jsimsa) [inline tests/src/test/java/alluxio/master/file/FileSystemMasterClientRestApiTest.java]: get rid off

File: core/server/src/main/java/alluxio/master/file/options/GetFileInfoListOptions.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
+/*
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/options/GetFileInfoListOptions.java]: add unit test for this class

**Comment by jsimsa** [comment]
@peisun1115 thank you for the new feature, left some comments for you

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java (hunk 1 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: Done.

File: core/common/src/thrift/common.thrift (hunk 2 above)
Comment (peisun1115) [inline core/common/src/thrift/common.thrift]: done

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 3 above)
Comment (peisun1115) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: done

File: core/server/src/main/java/alluxio/master/file/FileSystemMasterClientRestServiceHandler.java (hunk 4 above)
Comment (peisun1115) [inline core/server/src/main/java/alluxio/master/file/FileSystemMasterClientRestServiceHandler.java]: OK. Then I have to set the default value of the loadMetadataType input as empty string to indicate that the param is not set.

Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/FileSystemMasterClientRestServiceHandler.java]: makes sense an option value should be optional

File: core/server/src/main/java/alluxio/master/file/FileSystemMasterClientRestServiceHandler.java (hunk 5 above)
Comment (peisun1115) [inline core/server/src/main/java/alluxio/master/file/FileSystemMasterClientRestServiceHandler.java]: done

File: tests/src/test/java/alluxio/master/file/FileSystemMasterClientRestApiTest.java (hunk 6 above)
Comment (peisun1115) [inline tests/src/test/java/alluxio/master/file/FileSystemMasterClientRestApiTest.java]: done

File: core/server/src/main/java/alluxio/master/file/options/GetFileInfoListOptions.java (hunk 7 above)
Comment (peisun1115) [inline core/server/src/main/java/alluxio/master/file/options/GetFileInfoListOptions.java]: done

**Comment by peisun1115** [comment]
Jenkins, test this please

**Comment by peisun1115** [comment]
@jsimsa  I have addressed your comments. Thanks for your review.

File: core/common/src/main/java/alluxio/wire/LoadMetadataType.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+import alluxio.thrift.LoadMetadataTypeT;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Types for loading metadata.
+ */
+@ThreadSafe
+public enum LoadMetadataType {
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/wire/LoadMetadataType.java]: add `@PublicApi` annotation

File: core/server/src/main/java/alluxio/master/file/FileSystemMasterClientRestServiceHandler.java [hunk 9]
```diff
@@ -251,21 +254,32 @@ public Response free(@QueryParam("path") String path,
   }
 
   /**
-   * @summary get the file descriptors for a path
+   * @summary get the file descriptors for a path.
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/FileSystemMasterClientRestServiceHandler.java]: style: either `Get the file descriptors for a path.` or `get the file descriptors for a path`

**Comment by jsimsa** [comment]
two minor comments, otherwise LGTM

File: core/common/src/main/java/alluxio/wire/LoadMetadataType.java (hunk 8 above)
Comment (peisun1115) [inline core/common/src/main/java/alluxio/wire/LoadMetadataType.java]: done

File: core/server/src/main/java/alluxio/master/file/FileSystemMasterClientRestServiceHandler.java (hunk 9 above)
Comment (peisun1115) [inline core/server/src/main/java/alluxio/master/file/FileSystemMasterClientRestServiceHandler.java]: done

**Comment by peisun1115** [comment]
@jsimsa I have addressed your comments

**Comment by jsimsa** [comment]
LGTM

**Comment by peisun1115** [comment]
Jenkins, test this please

File: core/common/src/thrift/file_system_master.thrift [hunk 10]
```diff
@@ -25,8 +25,16 @@ struct MountTOptions {
   2: optional map<string, string> properties
 }
 
+enum LoadMetadataTypeT {
```
Comment (apc999) [inline core/common/src/thrift/file_system_master.thrift]: should we call this `LoadMetadataTType` to be consistent with `XXXTOptions`?

Comment (peisun1115) [inline core/common/src/thrift/file_system_master.thrift]: I don't know. Is the convention to put T before the last word? That sounds werid. But I can change it if you prefer.

**Comment by peisun1115** [comment]
Jenkins, test this please

File: core/common/src/main/java/alluxio/wire/LoadMetadataType.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Types for loading metadata.
+ */
+@PublicApi
+@ThreadSafe
+public enum LoadMetadataType {
+  /**
+   *  Never loads metadata.
```
Comment (apc999) [inline core/common/src/main/java/alluxio/wire/LoadMetadataType.java]: NIT: you have an extra space before "Never"

Comment (peisun1115) [inline core/common/src/main/java/alluxio/wire/LoadMetadataType.java]: Done

File: core/common/src/thrift/file_system_master.thrift (hunk 10 above)
Comment (peisun1115) [inline core/common/src/thrift/file_system_master.thrift]: done

**Comment by peisun1115** [comment]
@apc999 The tests passed locally. mvn test

**Comment by apc999** [comment]
LGTM

**Pull request merged**